### PR TITLE
fix: OS updates UX improvements — merged packages view, hold/unhold, modal accessibility

### LINKED ISSUE #18: OS Updates Gear: Bug Fixes, UX Improvements, and Comprehensive Test Suite
## Overview

The OS Updates gear has multiple bugs, missing visual feedback, and no test coverage. This issue tracks all fixes, UX improvements, and creation of a comprehensive test suite.

### 1. Duplicate Error Messages in Toast Notifications
**Problem:** Error toasts display messages like `"Failed to check for updates: Failed to check for updates"` — the action description is repeated twice.

**Root Cause:** The JavaScript error handling pattern creates an `Error` with a message like `"Failed to check for updates"` and then the `showToast()` call prepends the same prefix again:
```js
// In checkForUpdates():
if (!response.ok) throw new Error('Failed to check for updates');  // <-- sets err.message
// ...
showToast('Failed to check for updates: ' + err.message, 'error');  // <-- duplicates it
```

This pattern exists in multiple functions across `os-updates-page.js`.

**Fix:** Either:
- Change `throw new Error(...)` to throw with just the server response text (not the prefix), OR
- Remove the prefix from `showToast()` and just use `err.message` directly

**Affected functions (all in `os-updates-page.js`):**
- [ ] `checkForUpdates()` — line 139/147
- [ ] `installAllUpdates()` — line 170/180
- [ ] `installSecurityUpdates()` — line 200/211
- [ ] `installSelectedPackages()` — line 238/249
- [ ] `createSnapshot()` — line 312/316
- [ ] `restoreSnapshot()` — line 333/338
- [ ] `deleteSnapshot()` — line 354/358
- [ ] `doReboot()` — line 409/413
- [ ] `confirmPipxInstall()` — line 443/448
- [ ] `upgradePipxPackage()` — line 458/463
- [ ] `upgradeAllPipx()` — line 473/477
- [ ] `uninstallPipxPackage()` — line 494/498
- [ ] Search for same pattern in the templ file and any other JS files

### 2. "Check for Updates" Button Throws Error
**Problem:** Clicking "Check for Updates" shows `"Failed to check for updates: Failed to check for updates"` error toast.

**Investigation needed:**
- [ ] Verify the dashboard handler (`APITriggerUpdateCheckHandler`) correctly proxies to agent
- [ ] Verify agent handler (`handleTriggerCheck`) correctly runs `apt update`
- [ ] Check if the issue is a timeout (apt update can take >30s, but client timeout is 30s)
- [ ] Check if the agent returns plain text error vs JSON (dashboard `WriteHTTPError` uses `http.Error()` which is plain text, but JS may expect JSON)
- [ ] The JS doesn't pass `?stream=true` in the fetch call, so it uses sync mode which blocks until apt finishes — this likely times out on the 30s default client timeout
- [ ] Fix: Either increase timeout for update check operations, OR switch to streaming mode in the JS

### 3. "Upgrade All" Pipx Packages Error
**Problem:** Clicking "Upgrade All" on pipx packages throws an error.

**Investigation needed:**
- [ ] The JS `upgradeAllPipx()` sends `POST /api/os-updates/pipx/upgrade` with empty body `{}`
- [ ] Dashboard handler `APIPipxUpgradeHandler` decodes body — if `name` is empty, calls `client.UpgradeAllPipxPackages()`
- [ ] But `UpgradeAllPipxPackages()` calls `/api/v1/system/pipx/upgrade-all` on agent
- [ ] Verify the route exists on agent (`handlePipxUpgradeAll`) and `pipx upgrade-all` works
- [ ] Check if pipx is available on the target server

### 4. Snapshots: Restore Does Nothing
**Problem:** Clicking Restore shows confirmation dialog, pressing Restore closes it, but nothing visibly happens.

**Root cause candidates:**
- [ ] Check if `RestoreSnapshot` agent API actually performs the dpkg restore
- [ ] The JS calls the API, gets success, shows toast, then reloads after 3s — but there's no visual indicator of the restore being in progress
- [ ] Fix: Add a progress/loading indicator during restore, show detailed output, add snapshot status indicator showing which snapshot the system is currently on

### 5. Snapshots: Delete Does Nothing
**Problem:** Deleting a snapshot does nothing visible.

**Investigation needed:**
- [ ] Verify `DeleteSnapshot` agent API works (file deletion in `/var/lib/gearbox-agent/snapshots/`)
- [ ] Check if the `snapshotID` is passed correctly (JS uses `btn.dataset.snapshotId`)
- [ ] Dashboard uses `r.PathValue("id")` — verify route matches `{id}` path parameter
- [ ] Check the DELETE route registration on dashboard side

### 6. Update History Shows Nothing
**Problem:** Update history view is empty even though updates were manually installed via SSH.

**Investigation needed:**
- [ ] Agent reads `/var/log/apt/history.log` — verify this file exists and has content on target server
- [ ] Check if the parser handles all apt history formats (rotated logs, different Debian/Ubuntu formats)
- [ ] The parser may need to read rotated logs (`/var/log/apt/history.log.1.gz`, etc.)
- [ ] Also read `/var/log/dpkg.log` as a fallback/supplement
- [ ] Fix: Parse both apt history.log and dpkg.log, handle rotated logs

### 7. Remove Redundant Refresh Button
**Problem:** There's both a "Check for Updates" button and a refresh button right next to each other — they're redundant.

**Fix:**
- [ ] Remove the standalone refresh button from the header
- [ ] Make "Check for Updates" the single action that refreshes all data
- [ ] Auto-run update check every time the page loads (call `checkForUpdates()` on page load, or at minimum refresh the data from the agent)

### 8. Install Selected: Add Per-Package Visual Indicators
**Problem:** When clicking "Install Selected", there's no visual feedback that anything is happening in the package table.

**Current state:** The code actually HAS progress tracking infrastructure (`startInstallProgress`, `setCurrentPackageSpinner`, `markPackageCompleted`, etc.) but it may not be wired up correctly to the SSE events, or the `.package-status-cell` elements may not exist in the template.

**Fix:**
- [ ] Verify `.package-status-cell` exists in each package row in the template
- [ ] Verify SSE events (`apt.output`, `apt.completed`, `apt.failed`) are properly dispatched
- [ ] Add spinner icon next to package name when installation starts
- [ ] Change spinner to green checkmark on success, red X on failure
- [ ] Make status icons clickable to show a dialog with the install output log for that package
- [ ] Capture and store installation logs for each operation

### 9. Update Logs Viewer
**Problem:** No way to review update operation logs after the terminal modal is closed.

**Implementation:**
- [ ] Create an Update Logs section/modal accessible via a button in the page header
- [ ] Store operation logs persistently (agent-side, in `/var/lib/gearbox-agent/update-logs/`)
- [ ] Log viewer features: search, filter by date/operation, copy to clipboard, download as file
- [ ] Style similar to the main Logs gear page
- [ ] Each log entry should include: timestamp, operation type, packages involved, success/failure, full terminal output

### 10. Snapshot Status Indicator
**Problem:** No visual indicator showing which snapshot the OS is currently on (if any).

**Fix:**
- [ ] Add a "Current" badge/indicator to the snapshot that was most recently restored
- [ ] Track restore state in the snapshot metadata
- [ ] Show "Active" state on the snapshot row when it's the current system state

### 11. Comprehensive Test Coverage
**Problem:** Zero test files exist for the OS Updates gear.

**Requirements:**
- Target as close to 100% coverage as feasible
- Tests must be REAL — testing actual logic, not trivial pass-throughs
- Tests must cover real-world scenarios and edge cases
- A test server will be available for e2e tests

#### Unit Tests (Go)

**Agent side (`gearbox-agent/internal/gears/updates/`):**
- [ ] `updates_test.go` — Test `UpdatesCollector` methods:
  - `CheckUpdates()` — parsing apt output for available updates
  - `ListUpgradable()` — parsing `apt list --upgradable` output
  - `GetUpdateHistory()` — parsing `/var/log/apt/history.log` with various formats
  - `CreateSnapshot()` / `ListSnapshots()` / `DeleteSnapshot()` — file operations
  - `RestoreSnapshot()` — dpkg restore flow
  - `IsPipxAvailable()` — command detection
  - `ListPipxPackages()` — parsing pipx list output
  - `GetRebootRequiredPackages()` — parsing reboot-required files
  - `GetUnattendedUpgradesConfig()` — config file parsing
  - Edge cases: empty outputs, malformed data, missing files, permission errors
- [ ] `plugin_test.go` — Test HTTP handlers:
  - All API endpoint handlers with mock collector
  - Request validation (missing fields, bad JSON)
  - Response format verification
  - Error response formatting
  - Streaming vs sync mode branching
- [ ] `apt_runner_test.go` — Test streaming apt operations:
  - Operation lifecycle (start, progress, complete, fail, cancel)
  - Event emission
  - Concurrent operation handling

**Dashboard side (`gearbox/internal/gears/os_updates/`):**
- [ ] `handlers_test.go` — Test page handler with mock dependencies
- [ ] `plugin_test.go` — Test gear registration and lifecycle

**Dashboard handler (`gearbox/internal/framework/handler/`):**
- [ ] `os_updates_test.go` — Test proxy handlers:
  - Permission checks
  - Request forwarding to agent client
  - Error handling and response formatting
  - Audit logging

#### Functional Tests

- [ ] Integration tests with real agent client against test server
  - Full update check cycle
  - Package installation flow
  - Snapshot create/restore/delete cycle
  - Pipx operations
  - History retrieval

#### End-to-End Tests

**Framework:** Choose appropriate e2e framework (Playwright or similar) for the web UI.

- [ ] Research and select e2e test framework compatible with Go/templ/HTMX stack
- [ ] `e2e/os_updates_test.go` or `e2e/os-updates.spec.ts`:
  - Navigate to OS Updates page, verify data loads
  - Check for updates flow (button click → spinner → results)
  - Select packages and install (verify progress indicators)
  - Create snapshot → verify it appears in list
  - Install updates → verify packages marked as updated
  - Restore snapshot → verify rollback
  - Delete snapshot → verify removal
  - View update history
  - Pipx operations (install, upgrade, uninstall)
  - Error handling (agent offline, permission denied)
  - Multi-server switching

#### Test Infrastructure
- [ ] Create test helpers/fixtures for common test setup
- [ ] Mock implementations for agent client, database, auth
- [ ] Test data generators for packages, snapshots, history entries
- [ ] CI integration (add test commands to Makefile)
- [ ] Update CLAUDE.md with test patterns and testing goals

## Implementation Order

1. **Phase 1: Bug Fixes** — Items 1-6 (fix existing broken functionality)
2. **Phase 2: UX Improvements** — Items 7-10 (improve user experience)
3. **Phase 3: Test Suite** — Item 11 (comprehensive testing)

### Dashboard (`gearbox/`)
- `internal/framework/handler/os_updates.go` — API proxy handlers
- `internal/framework/templates/pages/os_updates.templ` — Page template
- `static/js/os-updates/os-updates-page.js` — Client-side logic
- `internal/gears/os_updates/handlers.go` — Gear page handler
- `internal/framework/agent/client.go` — Agent client methods
- `internal/framework/errors/errors.go` — Error formatting

### Agent (`gearbox-agent/`)
- `internal/gears/updates/plugin.go` — API handlers
- `internal/gears/updates/updates.go` — Core collector logic
- `internal/gears/updates/apt_runner.go` — Streaming apt operations

### PR DESCRIPTION
**Pull request by sarg3nt**
## Summary

- **Merged OS Packages view**: Combined Available Updates and Installed Packages into a single section with an Updates/All Packages filter dropdown. Page load fetches only the upgradable list (fast); full installed list is lazy-loaded on demand.
- **Hold/Unhold packages**: Added held badge in the grid, Unhold action button, and full hold/unhold API support across agent and dashboard.
- **Available Version column**: Shows current version in muted text when up to date; green when an update exists.
- **Removed Revert feature**: Removed Revert button and all related code — older versions are not reliably available in standard apt repos. Use snapshots instead.
- **Modal accessibility**: Added Escape key handler for all modals app-wide (OS updates + HAProxy editor). Primary action/close button is focused when a modal opens. Enter key submits confirm modals.

## Test Plan

- [ ] OS Packages defaults to Updates view on page load (fast — only upgradable packages fetched)
- [ ] Switching to All Packages lazy-loads full list (spinner visible briefly)
- [ ] Switching back to Updates uses cached upgradable list (no re-fetch)
- [ ] Available Version shows green version when update exists; muted current version otherwise
- [ ] Hold a package — held badge appears, Unhold button replaces Remove in actions
- [ ] Unhold a package — badge disappears, normal state restored
- [ ] Press Escape to close any open modal
- [ ] Tab to buttons in modals; primary button is pre-focused; Enter activates it
- [ ] HAProxy editor: Escape closes backups, validation, marker-error, save-reason, and snippets modals

Closes #18

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gearbox/static/js/utils/toast.js [hunk 1]
```diff
@@ -1,174 +1,473 @@
...
+		});
+	}
+
+	// ── Enforce max toasts ─────────────────────────────────────────────
+
+	function enforceMaxToasts() {
+		while (activeToasts.length > DEFAULTS.maxToasts) {
+			// Remove the oldest toast (first in array)
+			dismiss(activeToasts[0]);
+		}
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox/static/js/utils/toast.js:326]: Fixed in 81ada81. enforceMaxToasts() now splices the toast from activeToasts synchronously before calling dismiss(), so the array length decreases each iteration and the loop terminates correctly.

File: gearbox-agent/internal/gears/updates/apt_runner.go [hunk 2]
```diff
@@ -389,3 +588,120 @@ func (r *AptRunner) ListOperations() []*AptOperation {
...
+	}
+
+	for _, entry := range entries {
+		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
+			continue
+		}
+
+		// Check if filename contains the ID prefix
+		if !strings.Contains(entry.Name(), id[:8]) {
+			continue
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/apt_runner.go:698]: Fixed in 81ada81. Added a length check at the top of GetUpdateLog — returns an error if id is shorter than 8 characters instead of panicking.

File: gearbox/static/js/utils/toast.test.html [hunk 3]
```diff
@@ -0,0 +1,399 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+<meta charset="UTF-8">
+<title>Toast Library Tests</title>
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox/static/js/utils/toast.test.html]: Fixed in 81ada81. Deleted toast.test.html from static/ — it was only used for manual local testing and should not be shipped in production.

File: gearbox-agent/deploy/gearbox-agent.service [hunk 4]
```diff
@@ -25,10 +26,11 @@ Restart=always
...
 
 # Security hardening
+# Note: ProtectSystem is intentionally not set to "strict" because gearbox-agent
+# needs broad system access for package management (apt install/upgrade),
+# service control, HAProxy config, certificate management, and system reboots.
 NoNewPrivileges=false
-ProtectSystem=strict
-ProtectHome=read-only
-ReadWritePaths=/var/lib/gearbox-agent /etc/haproxy /etc/letsencrypt
+ProtectHome=no
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/deploy/gearbox-agent.service]: Fixed in 81ada81. Restored ProtectHome=read-only, PrivateTmp=true, and ProtectKernelTunables=true. ProtectSystem=strict and NoNewPrivileges remain off because the agent needs write access to /etc (HAProxy config, certs) and /var (package management, data dir), and runs package manager commands as root. Added a comment in the unit file documenting why.

File: gearbox-agent/internal/gears/updates/apt_runner.go [hunk 5]
```diff
@@ -332,6 +395,142 @@
...
+
+// runRestoreSnapshot executes snapshot restore and streams output.
+func (r *AptRunner) runRestoreSnapshot(ctx context.Context, op *AptOperation, snapshotID string) {
+	defer r.completeOperation(op)
+
+	snapshotDir := "/var/lib/gearbox-agent/snapshots"
+	snapshotFile := fmt.Sprintf("%s/%s.selections", snapshotDir, snapshotID)
+
+	// Verify snapshot exists
+	if _, err := os.Stat(snapshotFile); err != nil {
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/apt_runner.go:446]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/apt_runner.go [hunk 6]
```diff
@@ -332,6 +395,142 @@
...
+		r.failOperation(op, fmt.Sprintf("snapshot not found: %s", snapshotID))
+		return
+	}
+
+	// Read snapshot selections
+	selectionsData, err := os.ReadFile(snapshotFile) //#nosec G304
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/apt_runner.go:452]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/pm_apt.go [hunk 7]
```diff
@@ -0,0 +1,639 @@
...
+		return nil, fmt.Errorf("failed to save snapshot metadata: %w", err)
+	}
+	return &AptSnapshot{ID: snapshotID, CreatedAt: timestamp, Reason: reason}, nil
+}
+
+func (a *aptPackageManager) RestoreSnapshot(snapshotID string) error {
+	snapshotDir := "/var/lib/gearbox-agent/snapshots"
+	snapshotFile := fmt.Sprintf("%s/%s.selections", snapshotDir, snapshotID)
+
+	if _, err := os.Stat(snapshotFile); err != nil {
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/pm_apt.go:491]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/pm_apt.go [hunk 8]
```diff
@@ -0,0 +1,639 @@
...
+		return fmt.Errorf("snapshot not found: %s", snapshotID)
+	}
+	selectionsData, err := os.ReadFile(snapshotFile) //#nosec G304
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/pm_apt.go:494]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 9]
```diff
@@ -539,188 +353,257 @@
...
+}
 
-	// Verify snapshot exists
-	if _, err := os.Stat(snapshotFile); err != nil {
-		return fmt.Errorf("snapshot not found: %s", snapshotID)
+// computeDowngrades compares a snapshot's .versions file against currently
+// installed packages and returns a list of "package=version" strings for
+// packages that need to be downgraded.
+func computeDowngrades(versionsFile string) []string {
+	data, err := os.ReadFile(versionsFile) //#nosec G304
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:397]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 10]
```diff
@@ -539,188 +353,257 @@
...
 
-	// Remove both files
-	os.Remove(fmt.Sprintf("%s/%s.selections", snapshotDir, snapshotID))
-	os.Remove(fmt.Sprintf("%s/%s.meta", snapshotDir, snapshotID))
+	selectionsPath := fmt.Sprintf("%s/%s.selections", snapshotDir, snapshotID)
+	metaPath := fmt.Sprintf("%s/%s.meta", snapshotDir, snapshotID)
+	versionsPath := fmt.Sprintf("%s/%s.versions", snapshotDir, snapshotID)
+
+	// Check that the snapshot exists before attempting removal
+	_, selErr := os.Stat(selectionsPath)
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:457]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 11]
```diff
@@ -539,188 +353,257 @@
...
+	_, metaErr := os.Stat(metaPath)
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:458]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 12]
```diff
@@ -539,188 +353,257 @@
...
+	if os.IsNotExist(selErr) && os.IsNotExist(metaErr) {
+		return fmt.Errorf("snapshot not found: %s", snapshotID)
+	}
+
+	// Remove all snapshot files, ignoring errors for files that don't exist
+	os.Remove(selectionsPath)
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:464]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 13]
```diff
@@ -539,188 +353,257 @@
...
+	os.Remove(metaPath)
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:465]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 14]
```diff
@@ -539,188 +353,257 @@
...
+	os.Remove(versionsPath)
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:466]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 15]
```diff
@@ -539,188 +353,257 @@
...
-	logFile := "/var/log/apt/history.log"
-	file, err := os.Open(logFile)
+// PreviewRestore computes what changes restoring a snapshot would make without applying them.
+func (c *UpdatesCollector) PreviewRestore(snapshotID string) (*SnapshotPreview, error) {
+	snapshotDir := "/var/lib/gearbox-agent/snapshots"
+	selectionsFile := fmt.Sprintf("%s/%s.selections", snapshotDir, snapshotID)
+	versionsFile := fmt.Sprintf("%s/%s.versions", snapshotDir, snapshotID)
+
+	// Verify snapshot exists
+	if _, err := os.Stat(selectionsFile); err != nil {
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:498]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 16]
```diff
@@ -539,188 +353,257 @@
...
+		return nil, fmt.Errorf("snapshot not found: %s", snapshotID)
+	}
+
+	preview := &SnapshotPreview{
+		SnapshotID: snapshotID,
+	}
+
+	// Compare selections to find installs/removals
+	snapshotSelections, err := os.ReadFile(selectionsFile) //#nosec G304
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:507]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 17]
```diff
@@ -539,188 +353,257 @@
...
+		if exists && (snapState == "deinstall" || snapState == "purge") {
+			preview.Removals = append(preview.Removals, pkg)
+		}
 	}
 
-	// Apply limit
-	if limit > 0 && len(history) > limit {
-		history = history[:limit]
+	// Check for version-aware downgrades
+	if _, err := os.Stat(versionsFile); err == nil {
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:557]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/updates.go [hunk 18]
```diff
@@ -539,188 +353,257 @@
...
+		preview.HasVersions = true
+
+		// Read snapshot versions
+		versionsData, err := os.ReadFile(versionsFile) //#nosec G304
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/updates.go:561]: Fixed in 81ada81. Added validateSnapshotID() which rejects any snapshot ID containing path separators, dots, or other unsafe characters (only alphanumeric, hyphens, and underscores are allowed). This validation is called in StartRestore (apt_runner.go), RestoreSnapshot (pm_apt.go), DeleteSnapshot, and PreviewRestore (updates.go) before the ID is used to construct any file path.

File: gearbox-agent/internal/gears/updates/pm_apt.go [hunk 19]
```diff
@@ -0,0 +1,642 @@
...
+}
+
+
+// HoldPackage marks a package as held using apt-mark, preventing upgrades.
+func (a *aptPackageManager) HoldPackage(name string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	cmd := exec.CommandContext(ctx, "apt-mark", "hold", name)
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("apt-mark hold %s: %w: %s", name, err, strings.TrimSpace(string(out)))
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/pm_apt.go:613]: Fixed in 2c74c2c — added `isValidPackageName()` validation at the top of both `HoldPackage` and `UnholdPackage`, consistent with the existing pattern in `InstallPackage` and `RemovePackage`.

File: gearbox-agent/internal/gears/updates/detector.go [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+func (u *unsupportedPM) BuildInstallCommand(_ []string, _ bool) []string { return nil }
+func (u *unsupportedPM) BuildUpgradeCommand() []string                   { return nil }
+func (u *unsupportedPM) BuildUpdateCheckCommand() []string               { return nil }
+func (u *unsupportedPM) EnvVars() []string                               { return nil }
+
+func (u *unsupportedPM) HoldPackage(_ string) error                              { return errNotSupported(u.name) }
+func (u *unsupportedPM) UnholdPackage(_ string) error                            { return errNotSupported(u.name) }
+func (u *unsupportedPM) ListHeldPackages() ([]string, error)                     { return nil, nil }
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/detector.go]: Fixed in 2c74c2c — `unsupportedPM.ListHeldPackages()` now returns `errNotSupported(u.name)` instead of `(nil, nil)`, consistent with all other stubs on that type.

File: gearbox-agent/internal/gears/updates/apt_runner.go [hunk 21]
```diff
@@ -29,18 +35,38 @@ type AptOperation struct {
...
 	operations map[string]*AptOperation
+	logMu      sync.Mutex
+	logBuffers map[string]*strings.Builder
 }
 
 // NewAptRunner creates a new apt runner.
 func NewAptRunner(eventBus *events.Bus) *AptRunner {
+	// Ensure log directory exists
+	_ = os.MkdirAll(updateLogsDir, 0750)
+
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox-agent/internal/gears/updates/apt_runner.go:68]: Fixed in 2c74c2c — `NewAptRunner` now logs a warning via `slog.Warn` when `MkdirAll` fails, so the failure is visible in the agent logs instead of being silently swallowed.

File: gearbox/static/js/utils/char-limit.js [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+    counter.style.lineHeight = '1';
+
+    // Ensure the parent has relative positioning for the counter
+    const parent = el.parentElement;
+    if (parent && getComputedStyle(parent).position === 'static') {
+        parent.style.position = 'relative';
+    }
+    // Insert counter after the input
+    el.insertAdjacentElement('afterend', counter);
+
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox/static/js/utils/char-limit.js]: Fixed in 2c74c2c — `initCharLimit` now returns early when `el.parentElement` is null, preventing the `insertAdjacentElement` call from throwing on disconnected elements. The docstring note about dynamic elements still applies, but callers must ensure the element is in the DOM before calling.

File: gearbox/static/js/utils/char-limit.js [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+    // Create the counter element
+    const counter = document.createElement('span');
+    counter.className = 'char-limit-counter text-xs text-gray-400 dark:text-gray-500 absolute right-2 bottom-1 pointer-events-none select-none hidden';
+    counter.style.lineHeight = '1';
+
+    // Ensure the parent has relative positioning for the counter.
+    // Guard against elements not yet inserted into the DOM (no parentElement).
+    const parent = el.parentElement;
+    if (!parent) return;
+    if (getComputedStyle(parent).position === 'static') {
```
[the remark this answers is not in the document]
Comment (sarg3nt) [reply gearbox/static/js/utils/char-limit.js:31]: Fixed in 8d610f1 — moved `el._charLimitInit = true` to after the `!parent` guard, so elements called before DOM insertion are not permanently marked as initialized and `initCharLimit()` can be retried successfully after the element is inserted.
